src/__tests__/e2e/completion.test.ts

```diff
@@ -0,0 +1,74 @@
+import { afterEach, beforeEach, describe, expect, it } from "bun:test";
+import { CompletionList, type CompletionItem } from "vscode-languageserver-protocol";
+import { LSPTestClient } from "./lsp-client";
+
+describe("Completion", () => {
+  let client: LSPTestClient;
+
+  const isCompletionList = (result: any): result is CompletionList => {
+    return result.items !== undefined;
+  };
+
+  const extractLabels = (result: CompletionList | CompletionItem[]) => {
+    if (isCompletionList(result)) {
+      return result.items.map(item => item.label);
+    } else {
+      return result.map(item => item.label);
+    }
+  }
+
+  beforeEach(async () => {
+    client = new LSPTestClient();
+    await client.start();
+  });
+
+  afterEach(async () => {
+    try {
+      await client.shutdown();
+      await client.exit();
+    } catch (error) {
+      // Server might already be stopped
+    }
+    await client.stop();
+  });
```

Comment:
The equivalent to what we're doing in pytest would be beforeAll/afterAll, I think. Its a tradeoff between test encapsulation and speed -- it makes the tests a ton faster to only do the client setup/teardown once per module, but then we have to be extra careful about our tests not messing with each other. Food for thought

Reply:
Hmm, yeah. Let me think about that.